disttask/loaddata/scheduler.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/pingcap/tidb/disttask/framework/scheduler"
+	"github.com/pingcap/tidb/executor/importer"
+	"github.com/pingcap/tidb/table/tables"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+// ImportScheduler is a scheduler for load data.
+type ImportScheduler struct {
+	taskMeta      *TaskMeta
+	indexEngine   *backend.OpenedEngine
+	dataEngines   []*backend.OpenedEngine
+	tableImporter *importer.TableImporter
+}
+
+// InitSubtaskExecEnv implements the Scheduler.InitSubtaskExecEnv interface.
+func (s *ImportScheduler) InitSubtaskExecEnv(ctx context.Context) error {
+	logutil.BgLogger().Info("InitSubtaskExecEnv", zap.Any("taskMeta", s.taskMeta))
+
+	idAlloc := kv.NewPanickingAllocators(0)
+	tbl, err := tables.TableFromMeta(idAlloc, s.taskMeta.Plan.TableInfo)
+	if err != nil {
+		return err
+	}
+	controller, err := importer.NewLoadDataController(&s.taskMeta.Plan, tbl)
+	if err != nil {
+		return err
+	}
+	if err := controller.InitDataFiles(ctx); err != nil {
+		return err
+	}
+
+	tableImporter, err := importer.NewTableImporter(&importer.JobImportParam{
+		GroupCtx: ctx,
+	}, controller)
+	if err != nil {
+		return err
+	}
+	s.tableImporter = tableImporter
+
+	indexEngine, err := s.tableImporter.OpenIndexEngine(ctx)
+	if err != nil {
+		return err
+	}
+	s.indexEngine = indexEngine
+	return nil
+}
+
+// SplitSubtask implements the Scheduler.SplitSubtask interface.
+func (s *ImportScheduler) SplitSubtask(ctx context.Context, bs []byte) ([]proto.MinimalTask, error) {
+	logutil.BgLogger().Info("SplitSubtask", zap.Any("taskMeta", s.taskMeta))
+	var subtaskMeta SubtaskMeta
+	err := json.Unmarshal(bs, &subtaskMeta)
+	if err != nil {
+		return nil, err
+	}
+
+	dataEngine, err := s.tableImporter.OpenDataEngine(ctx, subtaskMeta.ID)
+	if err != nil {
+		return nil, err
+	}
+	s.dataEngines = append(s.dataEngines, dataEngine)
+
+	miniTask := make([]proto.MinimalTask, 0, len(subtaskMeta.Chunks))
+	for _, chunk := range subtaskMeta.Chunks {
+		miniTask = append(miniTask, MinimalTaskMeta{
+			Plan:          s.taskMeta.Plan,
+			Chunk:         chunk,
+			DataEngine:    dataEngine,
+			IndexEngine:   s.indexEngine,
+			TableImporter: s.tableImporter,
+		})
+	}
+	return miniTask, nil
+}
+
+// CleanupSubtaskExecEnv implements the Scheduler.CleanupSubtaskExecEnv interface.
+func (s *ImportScheduler) CleanupSubtaskExecEnv(ctx context.Context) (err error) {
+	defer func() {
+		err2 := s.tableImporter.Close()
+		if err == nil {
+			err = err2
+		}
+	}()
+
+	logutil.BgLogger().Info("CleanupSubtaskExecEnv", zap.Any("taskMeta", s.taskMeta))
+	// TODO: add OnSubtaskFinish callback in scheduler framework, so we can import immediately after subtask finished.
+	// TODO: If subtask failed, we should not import data.
+	for _, engine := range s.dataEngines {
+		closedEngine, err := engine.Close(ctx)
+		if err != nil {
+			return err
+		}
+		if err := s.tableImporter.ImportAndCleanup(ctx, closedEngine); err != nil {
```

Comment:
no need to change in this PR.

the semantics of `Cleanup` does not make me expect that it will do some real work. Maybe a better name is MinimalTaskJoin/Wait or something.

And still we can ask the questions from dataflow engine framework: What's the behaviour if `MinimalTaskJoin` / `CleanupSubtaskExecEnv `fails? will framekwork retry later and where is the starting point of retry? will some resources leak if framework don't retry?

Reply:
That's true. I prepare to add a new callback(OnSubtaskFinish) in distframework to do post process